### ISSUE TITLE
Kubernetes HA

### ISSUE THREAD

**Issue by peterlamar** [CiscoCloud/kubernetes-ansible#15]
As a tenant, I can deploy Kubernetes in an HA manner so that deployments are more reliable and production ready

**Comment by kenjones-cisco** [comment]
Currently the only limitation I've seen was around the control where only a single instance of that works within Kubernetes. I believe that was related to api-server specifically.

**Comment by altvnk** [comment]
ref: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/admin/high-availability.md

**Comment by altvnk** [comment]
I think we can make something similar.

**Comment by kenjones-cisco** [comment]
So the limitation was a result of the original ansible than an inherent limitation to Kubernetes. I agree that should be doable!

**Comment by altvnk** [comment]
Should we split this into subtasks? For example we can start with `etcd` and make it more reliable.

**Comment by kenjones-cisco** [comment]
That would make sense instead of one big single change, and etcd is definitely the low hanging fruit to start with.

**Comment by peterlamar** [comment]
Sounds good to me, this is certainly an 'epic' level task. Feel free to create new issues and reference this one or just close this one.

**Comment by kenjones-cisco** [comment]
Based on the documentation from Kubernetes on running it in an HA setup, it shows running the different services as actual pods on the master nodes. Such that only kubelet is required to be installed locally on the host.

Are there any objections to switching from local install of all the Kubernetes services to pods managed by kubelet?
@PeterLamar @altvnk

**Comment by peterlamar** [comment]
Seems like a good idea to me, anyone have any downsides? This allows one to use all the k8s apis to help manage the services. Out of curiosity, do the official k8s repos recommend any particular path?

**Comment by altvnk** [comment]
This solution in proposal status.
Packaging looks more mature for me, but at least we can try and abandon packaging at all. But we shouldn't  maintain containers with k8s components since it's a huge amount of additional work.

**Comment by kenjones-cisco** [comment]
Complete agree @altvnk 

An example from the provided pod definition files:

```
gcr.io/google_containers/kube-apiserver
```

The images must be provided by Kubernetes directly otherwise packaging is the only approach.

**Comment by altvnk** [comment]
There is another cool thing that i saw - hyperkube (hypercube?) all-in-one binary for packaging in container. Contains all k8s components and runs needed by specifying argument at runtime.

**Comment by kenjones-cisco** [comment]
Great minds think alike, I had just been looking at the hyperkube container myself. A single container does simplify the number of image pulls.

**Comment by altvnk** [comment]
FYI:
- https://github.com/kubernetes/kubernetes/pull/13037
- https://github.com/kubernetes/kubernetes/pull/13040

**Comment by timothysc** [comment]
Is there a reason why you wouldn't just try to coordinate around upstream ansible here?https://github.com/kubernetes/contrib/tree/master/ansible .?.?.?

**Comment by altvnk** [comment]
There is no concerns to use the upstream one. In general the reason we started this is to provide the most easier way to automate platform installation that people asked for. So inspired by upstream automation we wanted to make it less complicated, plus we added some feature like Terraform. If we talking about HA, it's not available in k8s for the moment (including upstream ansible) and we looked for the way to make some PoC based on existing ideas.
